File: originbot_driver/kruisee_lidar/sdk_core/drivers/protocol.cc
//
// Copyright (c) 2021 ECOVACS
//
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT
//

#include "drivers/protocol.h"

#include <iostream>

#pragma pack(1)
// 不含能量信息的报文结构
struct MiniPacket {
    uint8_t header;
    uint8_t angle;
    uint16_t speed;
    uint16_t data[16];
    uint32_t stamp;
    uint16_t checksum;
};
#pragma pack()

bool Packet::IsHeader(uint8_t header)
{
    return header == 0xFA;
}

uint32_t Packet::length() { return sizeof(MiniPacket); }

// 该函数用于对数据包有效性校验(同时支持含能量和不含能量信息两种报文的解析)
bool Packet::IsValid(const Packet *packet, bool &energetic)
{
    // checksum angle
    if (packet->angle < 0xA0 || packet->angle > 0xF9)
        return false;

    // checksum minimum packet
    const MiniPacket *minipacket = reinterpret_cast<const MiniPacket *>(packet);
    const uint8_t *data = reinterpret_cast<const uint8_t *>(minipacket);

    uint32_t sum = 0;
    for (uint32_t i = 0; i < sizeof(*minipacket) - sizeof(packet->checksum); i++)
        sum += *(data + i);

    if ((sum == minipacket->checksum) && Packet::IsHeader((minipacket + 1)->header)) {
        energetic = false;
        return true;
    }

    for (uint32_t i = sizeof(*minipacket) - sizeof(minipacket->checksum); i < sizeof(*packet) - sizeof(packet->checksum); i++)
        sum += *(data + i);

    if (sum == packet->checksum) {
        energetic = true;
        return true;
    }

    return false;
}

File: originbot_driver/kruisee_lidar/sdk_core/config/protocol.lua
--
-- Copyright (c) 2022 ECOVACS
--
-- Use of this source code is governed by a MIT-style
-- license that can be found in the LICENSE file or at
-- https://opensource.org/licenses/MIT
--

-- 该文件负责激光雷达通讯相关配置

-- 串口通讯方式配置
UART = {
    -- 串口号
    name = "/dev/ttyUSB0",
    -- 波特率
    baudrate = 460800,
}

-- 以太网通讯方式配置
ETHERNET = {
    -- 激光雷达IP
    radar_ip = "169.254.119.2",
    -- 激光雷达端口
    radar_port = 3000,

    -- PC监听端口
    listen_port = 2000,
}

PROTOCOL = {
    uart = UART,
    ethernet = ETHERNET,

    use_net_protocol = false,
    use_uart_protocol = false
}